src/main/java/fr/sncf/osrd/train/TrainState.java

```diff
@@ -219,4 +229,35 @@ public void setAspectConstraints(SignalState signalState) {
         var navigatePhase = (SignalNavigatePhase.State) currentPhaseState;
         navigatePhase.setAspectConstraints(signalState, this);
     }
+
+    /** Occupy and free tvd sections given a detector the train is interacting with. */
+    public void updateTVDSections(
+            Simulation sim,
+            Detector detector,
+            InteractionType interactionType
+    ) throws SimulationError {
+        // Update route index
+        var currentRoute = path.routePath.get(routeIndex);
+        var tvdSectionPathIndex = currentRoute.tvdSectionsPaths.size() - 1;
+        var lastTvdSectionPath = currentRoute.tvdSectionsPaths.get(tvdSectionPathIndex);
+        var lastTvdSectionPathDir = currentRoute.tvdSectionsPathDirections.get(tvdSectionPathIndex);
+        if (lastTvdSectionPath.getEndNode(lastTvdSectionPathDir) == detector.index)
+            routeIndex++;
+
+        // Occupy the next tvdSection
+        if (interactionType == InteractionType.HEAD) {
+            var forwardTVDSectionPath = path.findForwardTVDSection(detector);
+            if (forwardTVDSectionPath == null)
+                return;
+            var nextTVDSection = sim.infraState.getTvdSectionState(forwardTVDSectionPath.index);
+            nextTVDSection.occupy(sim);
+            return;
+        }
```

Comment:
these two branches are symmetric, I'd rather like to see an if / else rather than a return